src/python/pants/backend/codegen/protobuf/target_types.py

```diff
@@ -128,6 +130,13 @@ class ProtobufSourcesGeneratorTarget(TargetFilesGenerator):
     settings_request_cls = GeneratorSettingsRequest
     help = "Generate a `protobuf_source` target for each file in the `sources` field."
 
+    @classmethod
+    def register_plugin_field(cls, field: Type[Field], move: bool = False) -> UnionRule:
+        rules = super().register_plugin_field(field)
+        if move and field not in cls.moved_fields:
+            cls.moved_fields = (*cls.moved_fields, field)
+        return rules
```

Comment:
Oh, you had such a long comment, I forgot this was a PR with code in it 😅 

Yeah I see we're thinking along the same lines. I'm however reluctant to introduce mutations on the class definition itself (reason being why that wasn't done that way already for the regular plugin fields).

Reply:
yes definitely, this is gruesome and goes against the existing design of having the extensibility come via Union lookup in the engine.